# Discrete_entropy, two dimensional array

Author: Heng Shi

This is discrete entropy assessment to see what in 2 factor system, how correlated sources affect the finally entropy.

### import libs

In [1]:
import numpy as np
import matplotlib as mp
import pandas as pd
from pandas import Series, DataFrame
import csv
import math

### define factors

In [2]:
input_name = './input_arr_independent.csv'
#input_name = './input_arr_correlated.csv'
df = pd.read_csv(input_name)
ds = np.array(df)

In [3]:
arr1 = ds[:,1]
arr2 = ds[:,2]
length = arr1.shape[0]
sce = max(arr1)
print length,sce

10000 9.0


### Proba 1D

In [4]:
prob1d_a = np.zeros((sce+1,1))
prob1d_b = np.zeros((sce+1,1))
for i in range(length):
    prob1d_a[arr1[i]] += 1
    prob1d_b[arr2[i]] += 1
prob1d_a = prob1d_a/(length*1.0)
prob1d_b = prob1d_b/(length*1.0)
print sum(prob1d_a)
print sum(prob1d_b)

[ 1.]
[ 1.]


/usr/local/lib/python2.7/dist-packages/ipykernel/__main__.py:1: DeprecationWarning: using a non-integer number instead of an integer will result in an error in the future
  if __name__ == '__main__':
/usr/local/lib/python2.7/dist-packages/ipykernel/__main__.py:2: DeprecationWarning: using a non-integer number instead of an integer will result in an error in the future
  from ipykernel import kernelapp as app
/usr/local/lib/python2.7/dist-packages/ipykernel/__main__.py:4: DeprecationWarning: using a non-integer number instead of an integer will result in an error in the future
/usr/local/lib/python2.7/dist-packages/ipykernel/__main__.py:5: DeprecationWarning: using a non-integer number instead of an integer will result in an error in the future


### prob2d

In [5]:
prob2d = np.zeros((sce+1,sce+1))
for i in range(length):
    prob2d[arr1[i]][arr2[i]] += 1
prob2d = prob2d/(length*1.0)
print prob2d
print sum(sum(prob2d))

[[ 0.011   0.0092  0.0107  0.0115  0.0104  0.0085  0.0112  0.0097  0.0095
   0.0088]
 [ 0.0104  0.0083  0.0125  0.0097  0.009   0.0106  0.01    0.0097  0.0107
   0.0083]
 [ 0.0103  0.0105  0.0099  0.0099  0.0092  0.0109  0.0089  0.0136  0.0112
   0.0095]
 [ 0.0086  0.01    0.0099  0.0107  0.011   0.0104  0.0108  0.0105  0.0093
   0.0098]
 [ 0.0099  0.0102  0.0106  0.0111  0.0092  0.0095  0.0112  0.0098  0.0109
   0.0102]
 [ 0.0097  0.0098  0.0089  0.0106  0.0091  0.009   0.0091  0.0116  0.0092
   0.0086]
 [ 0.0104  0.0091  0.0103  0.0093  0.0091  0.011   0.0093  0.01    0.0086
   0.0089]
 [ 0.0112  0.0085  0.0116  0.0108  0.0093  0.0093  0.0094  0.0082  0.0108
   0.01  ]
 [ 0.0107  0.0089  0.01    0.0111  0.0075  0.0113  0.0105  0.0118  0.0112
   0.0097]
 [ 0.0096  0.0092  0.0092  0.0101  0.0102  0.0099  0.0115  0.0104  0.0106
   0.0087]]
1.0


/usr/local/lib/python2.7/dist-packages/ipykernel/__main__.py:1: DeprecationWarning: using a non-integer number instead of an integer will result in an error in the future
  if __name__ == '__main__':
/usr/local/lib/python2.7/dist-packages/ipykernel/__main__.py:3: DeprecationWarning: using a non-integer number instead of an integer will result in an error in the future
  app.launch_new_instance()


### calculate entropy

In [6]:
def cal_ent(mat,dim):
    [sha1,sha2] = mat.shape
    len = sha1*sha2
    tmat = mat.reshape((len,1))
    ent_base = 0.0
    for i in range(len):
        if tmat[i] > 0:
            ent_base -= tmat[i]*np.log(tmat[i])
    return ent_base

### Assessment

In [7]:
# compare 1d 2d entropy
en_1d_a = cal_ent(prob1d_a,1)
print 'var a entropy: %2f' % en_1d_a
en_1d_b = cal_ent(prob1d_b,1)
print 'var b entropy: %2f' % en_1d_b
en_2d = cal_ent(prob2d,2)
print 'var a&b 2d entropy: %2f' % en_2d
en_disp = en_1d_a*1.0 + en_1d_b*1.0 - en_2d*1.0
print 'entropy disapeared: %2f' % en_disp
# correlation
cor = np.corrcoef(arr1,arr2)
print 'correlation coef: %2f' % cor[0,1]

var a entropy: 2.302246
var b entropy: 2.301540
var a&b 2d entropy: 4.600096
entropy disapeared: 0.003689
correlation coef: 0.009621


This result comes out a nature of randomness. For 2 independent variables, the combination of entropy is about to its summation

And 2 correlated variables are going to generate related entropy

Correlated results:

    var a entropy: 2.302089

    var b entropy: 2.113918

    var a&b 2d entropy: 3.977882

    entropy disapeared: 0.438125

    correlation coef: 0.706532

independent results:

    var a entropy: 2.302246

    var b entropy: 2.301540

    var a&b 2d entropy: 4.600096

    entropy disapeared: 0.003689

    correlation coef: 0.009621